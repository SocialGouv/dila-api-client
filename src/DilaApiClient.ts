import Debug from "debug";
import fetch from "node-fetch";
import OAuth2 from "simple-oauth2";

import { API_HOST, CREDENTIALS, TOKEN_HOST } from "./constants";

const debug = Debug("@socialgouv/dila-api-client");

export class DilaApiClient {
  public globalToken?: string;

  constructor(public apiHost = API_HOST, public tokenHost = TOKEN_HOST) {}

  public async getAccessToken(credentials = CREDENTIALS) {
    if (this.globalToken) {
      return this.globalToken;
    }
    const oauth2 = OAuth2.create(credentials);

    try {
      const result = await oauth2.clientCredentials.getToken({
        scope: "openid",
      });
      const accessToken = oauth2.accessToken.create(result);
      this.globalToken = accessToken.token.access_token;
    } catch (error) {
      debug("error", error);
      debug("Access Token error", error.message);
      throw error;
    }
    return this.globalToken;
  }

  public async fetch({
    path,
    method = "POST",
    params,
  }: {
    path: string;
    method?: string;
    params: object;
  }): Promise<any> {
    const [routeName] = path.split("/").slice(-1);
    const body = JSON.stringify(params);
    debug(`fetching route ${routeName} with ${body}...`);
    const token = await this.getAccessToken();
    const url = `${this.apiHost}/${path}`;

    const data = await fetch(url, {
      body,
      headers: {
        Authorization: `Bearer ${token}`,
        "content-type": "application/json",
      },
      method,
    }).then((r) => {
      if (r.status === 401 && this.globalToken) {
        this.globalToken = undefined;
        return this.fetch({ method, params, path });
      }
      return r.json();
    });

    if (data.error) {
      throw new Error(`Error on API fetch: ${JSON.stringify(data)}`);
    }

    return data;
  }
}
