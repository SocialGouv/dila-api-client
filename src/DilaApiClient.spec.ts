import fetchMock, { FetchMock } from "jest-fetch-mock";
import OAuth2 from "simple-oauth2";
import { DilaApiClient } from "./DilaApiClient";

jest.mock("simple-oauth2");

const fetch = fetchMock as FetchMock;
jest.mock("node-fetch", () => fetchMock);

beforeEach(() => {
  (OAuth2.create as jest.Mock).mockReset();
});

test("getAccessToken return token for OAuth2 transaction", async () => {
  (OAuth2.create as jest.Mock).mockReturnValue({
    accessToken: {
      create: () => ({
        token: { access_token: 1234 }
      })
    },
    clientCredentials: {
      getToken: () => Promise.resolve("fox")
    }
  });
  const client = new DilaApiClient();

  const token = await client.getAccessToken();

  expect(OAuth2.create).toHaveBeenCalledTimes(1);
  const [[credentials]] = (OAuth2.create as jest.Mock).mock.calls;
  expect(credentials).toMatchInlineSnapshot(`
    Object {
      "auth": Object {
        "authorizePath": "/api/oauth/authorize",
        "tokenHost": "https://oauth.aife.economie.gouv.fr",
        "tokenPath": "/api/oauth/token",
      },
      "client": Object {
        "id": "",
        "secret": "",
      },
      "options": Object {
        "authorizationMethod": "body",
      },
    }
  `);

  expect(token).toBe(1234);
  expect(client.globalToken).toBe(1234);
});

test("getAccessToken return in memory token", async () => {
  const client = new DilaApiClient();
  client.globalToken = "foo";

  const token = await client.getAccessToken();

  expect(OAuth2.create).not.toHaveBeenCalled();
  expect(token).toBe("foo");
  expect(client.globalToken).toBe("foo");
});

test("getAccessToken forward http error", async () => {
  (OAuth2.create as jest.Mock).mockReturnValue({
    clientCredentials: {
      getToken: () => Promise.reject(new Error("418 I'm a teapot"))
    }
  });
  const client = new DilaApiClient();

  expect(client.getAccessToken()).rejects.toThrowErrorMatchingInlineSnapshot(
    `"418 I'm a teapot"`
  );
  expect(client.globalToken).toBeUndefined();
});

//

test("fetch send correct API parameters", async () => {
  fetch.mockResponse(JSON.stringify({ message: "YATTA!" }));
  const client = new DilaApiClient();
  client.globalToken = "token";
  const res = await client.fetch({
    params: {
      date: 1234,
      sctId: "",
      textId: "LEGITEXT000006072050"
    },
    path: "consult/code/tableMatieres"
  });
  expect(res).toMatchInlineSnapshot(`
            Object {
              "message": "YATTA!",
            }
      `);
  expect(fetch.mock.calls.length).toEqual(1);
  const [[url, meta]] = fetch.mock.calls;
  expect(url).toEqual(`${client.apiHost}/consult/code/tableMatieres`);
  expect(meta).toMatchSnapshot();
});

test("fetch forward http error", async () => {
  fetch.mockReject(new Error("404"));
  const client = new DilaApiClient();
  client.globalToken = "token";
  expect(
    client.fetch({
      params: {},
      path: "whatever"
    })
  ).rejects.toThrowErrorMatchingInlineSnapshot(`"404"`);
});

test("fetch forward api error", async () => {
  fetch.mockResponse(JSON.stringify({ error: "YATTA!" }));
  const client = new DilaApiClient();
  client.globalToken = "token";
  expect(
    client.fetch({
      params: {},
      path: "whatever"
    })
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"Error on API fetch: {\\"error\\":\\"YATTA!\\"}"`
  );
});
