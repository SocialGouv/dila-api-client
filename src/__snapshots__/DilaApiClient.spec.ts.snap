// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch send correct API parameters 2`] = `
Object {
  "body": "{\\"date\\":1234,\\"sctId\\":\\"\\",\\"textId\\":\\"LEGITEXT000006072050\\"}",
  "headers": Object {
    "Authorization": "Bearer token",
    "content-type": "application/json",
  },
  "method": "POST",
}
`;
